code/lab/wmcgui.h

```diff
@@ -542,6 +542,7 @@ class Tree : public GUIObject
 	void ClearItems();
 
 	TreeItem* GetSelectedItem(){return SelectedItem;}
+	void SetSelectedItem(TreeItem *item) { SelectedItem = item; }
```

Comment:
This function changes the selected item but does not fire the listener function of that item. I would suggest that you move the relevant code from `Tree::DoMouseUp` into this function and then use this function in `Tree::DoMouseUp` to keep the functionality in one place.

Also, since the function will be quite big then it would be better to move the body into `wmcgui.cpp`.